[클린코드 7기 김진주] 영화 리뷰 미션 STEP 2

devpearlkim opened the pull request:
안녕하세요 윤호님,

요구사항에 따라 다음 기능을 구현했습니다.

### 구현 내용
- 검색 기능 추가
- 에러 상황 시 `alert` 알림 표시
- 타입스크립트로 전환
- 검색 관련 E2E 테스트 작성

[깃헙 페이지](https://devpearlkim.github.io/js-movie-review/)

요구사항을 따라 구현하는 데 집중했으며, 이번에는 특별히 궁금한 점은 없었습니다.

### 주요 변경 사항
- 기존에 `init`을 `Tab`에 통째로 넘기던 방식에서 `onTabChange`를 탭 선택 이벤트에 연결
- 컨트롤러에서 검색과 카테고리를 분리하고, `currentMode` 상태를 관리 (`currentMode`는 아직 활용되지 않음)
- `Header`를 한 번만 불러오도록 변경하여 불필요한 이벤트 리스너 중복 제거

감사합니다.

devpearlkim committed:
feat: Modularized HTML structure

devpearlkim committed:
feat: Fetch and display movie list from TMDB API

devpearlkim committed:
feat: Separate header, footer, and tabs components & add skeleton UI for movie list

devpearlkim committed:
feat: Implement "Load More" functionality and modularize movie list components

- Added "Load More" button to dynamically load additional movies
- Separated movie rendering logic into `movieRenderer.js`
- Moved "Load More" button creation/removal to `loadMoreButton.js`

devpearlkim committed:
chore: Delete custom button style

devpearlkim committed:
feat: Add responsive design & header background image

devpearlkim committed:
feat: Dynamically update header with first movie & add error message UI

devpearlkim committed:
chore: ensure previous movies persist when loading more

devpearlkim committed:
feat: Separate error handling and UI feedback for better user experience

devpearlkim committed:
chore: Setup Cypress for E2E testing

devpearlkim committed:
refactor: Improve Tabs component, add Header template, and optimize resize event

- Replaced innerHTML in Tabs with createElement for better security and performance
- Converted Header to use a template for cleaner separation of structure and logic
- Added debounce to resize event to optimize performance

devpearlkim committed:
refactor: Optimize DOM updates in movieRenderer.js using DocumentFragment

devpearlkim committed:
feat: Update movie fetching based on selected tab

devpearlkim committed:
test: Add skeleton UI and tab functionality tests

devpearlkim committed:
test: Add tests for data fetching and load more functionality

devpearlkim committed:
test: Add header rendering and error handling tests

devpearlkim committed:
chore: Delete unused file

devpearlkim committed:
refactor: Separate domain logic and UI with functional approach

devpearlkim committed:
feat: Add search functionality with URL state management

- Separated header logic into LoadBaseHeader (renders base header with search form) and updateHeaderMovie (updates movie info)
- Added search service logic to fetch search results via API
- Updated controller to attach search form submit event and handle search actions

devpearlkim committed:
refactor: Reorganize controller for mode management and event handling

- Moved base header loading to main for one-time initialization
- Modified controller to attach search form event only once
- Implemented mode state management ("search" vs "category") with separate functions (loadCategory, handleSearch)
- Added popstate handling for proper history state restoration
- Updated tabs to only trigger switchTab in the controller

devpearlkim committed:
refactor: migrate project to Typescript

devpearlkim committed:
fix: STEP1 피드백 반영

devpearlkim committed:
test: Switch error UI test to alert-based error messages

devpearlkim committed:
refactor: Replace DOM manipulation with template literals for better readability

devpearlkim committed:
test: Add E2E tests for search functionality and query handling

devpearlkim committed:
chore: Fix import path of constants

yujo11 commented on cypress/e2e/movies.cy.js:
```diff
@@ -132,4 +158,55 @@ describe("Movie App E2E Tests", () => {
     });
   });
+
+  describe("Search Functionality Tests", () => {
```
요거 저번 리뷰 때 여쭤보려다 깜박한 부분인데 영어로 테스트명을 작성하시는 이유가 있나용? (진짜 단순히 궁금해서 질문 드립니다 ㅎㅎ)

devpearlkim replied:
https://frontendmasters.com/courses/cypress/
제가 여기 인강으로 많이 공부하는데, 영어로 되어있어서 + 영어공부할겸 영어로 하고 있습니당,
일하면서는 거의 한글로 적습니다,,!

yujo11 commented on src/api.js:
```diff
@@ -32,3 +32,20 @@ export async function fetchMovies(category, page = 1) {
   }
 }
+
+export async function fetchSearchMovies(query, page = 1) {
+  const url = `${BASE_URL}/search/movie?language=ko-KR&page=${page}&query=${encodeURIComponent(
+    query
+  )}`;
+  const options = {
+    method: "GET",
+    headers: {
+      accept: "application/json",
+      Authorization: `Bearer ${API_KEY}`,
+    },
+  };
+  const response = await fetchWithTimeout(url, options, 10000);
+  if (!response.ok) throw new Error(`HTTP error status: ${response.status}`);
+  const data = await response.json();
+  return data.results;
+}
```
`fetch` 하는 함수들을 보면 반복적으로 사용되는 코드가 많이 보이는데 한 단계 추상화를 해보셔도 좋을거 같네용!

devpearlkim replied:
넵, 어제 실시간 수업에서도 강사님이 말씀해주셨던 부분이네요,,! 적용해보겠습니다 감사합니다!

yujo11 commented on src/components/MovieRenderer.ts:
```diff
@@ -0,0 +1,67 @@
+
+  movieContainer.innerHTML = "";
+
+  const skeletonItems = Array.from({ length: 8 })
+    .map(
+      () => `
+      <div class="movie-item skeleton">
+        <div class="item">
+          <div class="skeleton skeleton-thumbnail"></div>
+          <div class="item-desc">
+            <div class="skeleton-star-row">
+              <div class="skeleton skeleton-star"></div>
+              <p class="skeleton skeleton-rate"></p>
+            </div>
+            <div class="skeleton-title-row">
+              <strong class="skeleton skeleton-title"></strong>
+            </div>
+          </div>
+        </div>
+      </div>
+    `
+    )
+    .join("");
+
+  movieContainer.innerHTML = `<div class="movie-grid">${skeletonItems}</div>`;
```
코드를 보기 편하고 DOM에 한번만 접근하도록 신경써주신 부분이 좋네용 👍 👍 👍

yujo11 commented on src/controllers/createMovieController.ts:
```diff
@@ -0,0 +1,228 @@
+  async function fetchMoviesBySearch(query: string): Promise<void> {
+    if (!movieContainer) return;
+
+    currentMode = "search";
+    setMode("search");
+
+    showSkeletonUI(movieContainer);
+
+    try {
+      await service.searchMovies(query);
+
+      movieContainer.innerHTML = "";
+      renderNextBatch();
+
+      if (service.hasMore()) {
+        if (!loadMoreButtonComponent) {
+          loadMoreButtonComponent = LoadMoreButton(
+            movieContainer,
+            renderNextBatch
+          );
+        }
+        loadMoreButtonComponent.render();
+      } else if (loadMoreButtonComponent) {
+        loadMoreButtonComponent.remove();
+      }
+
+      if (!service.getFirstMovie()) {
+        movieContainer.innerHTML = "<p>검색 결과가 없습니다.</p>";
+      }
+
+      history.replaceState(
+        { query },
+        "",
+        `?search=${encodeURIComponent(query)}`
+      );
+    } catch (error) {
+      console.error("영화 검색 중 오류가 발생했습니다", error);
+      showErrorMessage("검색 중 오류가 발생했습니다.");
+    }
+  }
```
코드 블럭이 길어져 가독성이 다소 떨어지는데 적절한 단위로 함수를 나눈 다음에 호출해줘도 좋을거 같아요!

yujo11 commented on src/controllers/createMovieController.ts:
```diff
@@ -0,0 +1,228 @@
+    await fetchMoviesByCategory(newCategory);
+  }
+
+  async function fetchMoviesByCategory(category: Category): Promise<void> {
```
요 친구도 👆와 같은 맥락에서 로직을 나줘보셔도 좋을거 같네요

yujo11 commented on src/controllers/createMovieController.ts:
```diff
@@ -0,0 +1,228 @@
+    }
+  }
+
+  function setMode(mode: "category" | "search"): void {
+    currentMode = mode;
+    const tabContainer = document.getElementById(
+      "tab-container"
+    ) as HTMLElement | null;
+    if (mode === "search") {
+      if (tabContainer) {
+        tabContainer.style.display = "none";
+      }
+    } else {
+      if (tabContainer) {
+        tabContainer.style.display = "block";
+      }
+      const inputEl = document.querySelector(
+        ".search-input"
+      ) as HTMLInputElement | null;
+      if (inputEl) inputEl.value = "";
+    }
+  }
```
`if - esle` 가 여러 번 중첩되어 사용되면서 가독성이 떨어지고 있는데요. 개선시켜 볼 수 있을까요?

devpearlkim replied:
앗넵 감사합니다!

yujo11 commented on src/types/type.ts:
```diff
@@ -0,0 +1,29 @@
+export type Category = "now_playing" | "popular" | "top_rated" | "upcoming";
+
+export interface ApiMovie {
+  id: number;
+  title: string;
+  poster_path: string;
+  vote_average: number;
+  backdrop_path: string;
+}
+
+export interface Movie {
+  id: number;
+  title: string;
+  poster_path: string;
+  vote_average: number;
+  backdrop_path: string;
+  getThumbnailUrl: () => string;
+  getBackdropUrl: () => string;
+  getFormattedVote: () => string;
+}
+
+export interface MovieService {
+  loadMovies: (category: Category) => Promise<void>;
+  searchMovies: (query: string) => Promise<void>;
+  getNextBatch: () => Movie[];
+  hasMore: () => boolean;
+  getFirstMovie: () => Movie | null;
+  setMoviesPerLoad: (num: number) => void;
+}
```
type 명명 규칙이 있으실까요? 각각의 타입명만 보고는 사용처를 유추하기가 족므 어렵게 느껴져서용

devpearlkim replied:
작성한 규칙은 따로 없습니다, 한번 찾아보고 규칙을 적용해보겠습니다!

(수정)
찾아보니 보편적인 규칙이 있었네요, 타입스크립트는 사실상 거의 GPT에 의존해서 다 했는데 기본적인것은 공부해야겠군요,
 감사합니다

yujo11 commented on src/controllers/createMovieController.ts:
```diff
@@ -0,0 +1,228 @@
+  ) as HTMLElement | null;
+  const service: MovieService = createMovieService();
+
+  let currentMode: "category" | "search" = "category";
```
이 친구는 어떤 용도로 사용되고 있을까용?

devpearlkim replied:
지금 사용을 안하고 있습니다,,! 삭제하는게 좋을것같습니다, 감사합니다

(수정)
삭제했다가 검색이랑 카테고리fetch의 중복을 분리하는 과정에서 다시 추가하게 되었습니다,,!

currentMode는 현재 탭 전환 함수(switchTab)에서 검색 모드일 때 탭 전환을 막는 용도,
fetchMovies에서 query 여부를 직접 확인하는 대신 currentMode를 활용하여
검색 모드와 카테고리 모드를 구분하는 데 사용하고 있습니다.

(추가) 사실 계속 하다 보니까 currentMode를 관리하는 의미가 딱히 있는지는 모르겠습니다,
몇군데에서 써서 그냥 남겨두었습니다

yujo11 reviewed (changes requested):
안녕하세요 진주님 ~ 

몇몇 개선되면 좋을거 같은 부분들 코멘트로 남겼는데요!

코멘트 확인해주시고 다시 리뷰요청 부탁드립니다!! 고생하셨습니다!! 👍 👍 👍

devpearlkim committed:
chore: remove unused variable and api.js

devpearlkim committed:
refactor: Abstract common fetch logic into reusable fetchData helper

devpearlkim committed:
refactor: Update naming conventions for movie-relatited types

devpearlkim committed:
fix: Ensure non-null container assignment

devpearlkim committed:
refactor: Replace setMode with updateTabContainer and resetSearchInput function

devpearlkim committed:
refactor: Improve modularity and state management of MovieController

- Extract UI-related functions to `ui.ts`
- Introduce `currentMode` for better state handling
- Refactor `fetchMovies` logic for improved readability

devpearlkim committed:
refactor: Extract getCurrentMode

devpearlkim committed:
chore: Handle potential null return from createMovieController

devpearlkim committed:
refactor: Remove currentCategory and pass Tabs to controller init

devpearlkim committed:
chore: Move tabItems, validCategories, and defaultCategory to constants

yujo11 commented on src/services/api.ts:
```diff
@@ -11,58 +11,80 @@ function fetchWithTimeout(
+  init?: RequestInit;
+};
+
+async function fetchData<T>({
```
👍 👍 👍

yujo11 commented on src/types/type.ts:
```diff
@@ -19,11 +23,11 @@ export interface Movie {
 
-export interface MovieService {
-  loadMovies: (category: Category) => Promise<void>;
+export interface IMovieService {
```
여기만 `I-` prefix를 사용하는 이유가 있을까요? 어떤 명명 규칙을 따르셨는지 잘 모르곘네용

devpearlkim replied:
API에서 불러온거는 DTO를 붙이고 모델은 Model붙였는데 그거는 뭐라고 할지 몰라서 그냥 Interface라서  I붙였습니다,,! 이렇게 하는게 아니었군요, 좀 찾아보고 개선하겠습니다,,!

yujo11 replied:
아하 그렇군요 이런 헝가리안 표기법은 요즘 많이 사용되지 않는 방법인거 같아요.(쓰지 말라는 얘기는 아닙니다 !)

타입 네이밍 관련해서 명확한 룰 같은 것은 크게 본 적이 없는거 같은데요. 진주님의 코드베이스 내에서 일관성 있게 작성되도록만 신경 써주셔도 좋을거 같습니다 !

yujo11 commented on src/controllers/createMovieController.ts:
```diff
@@ -56,18 +61,18 @@ export function createMovieController(containerId: string) {
-      if (inputEl) inputEl.value = "";
+      if (inputEl) {
+        inputEl.value = "";
+      }
```
오 한결 보기 편해졌네요 👍 👍 👍

yujo11 commented on src/controllers/createMovieController.ts:
```diff
@@ -195,13 +175,27 @@ export function createMovieController(containerId: string) {
+    if (inputEl) {
+      inputEl.value = "";
+    }
+  }
```
👍 👍 👍

yujo11 reviewed (approved):
안녕하세요 진주님 ~~

함수들의 가독성이 많이 향상된 점이 눈에 띄네요 ! 하나의 함수에서 많은 일들을 하게 내부 분기가 많아지고 자연스럽게 코드를 읽기가 어려워지는데요. 하나의 함수가 한 가지 역할만 할 수 있도록 앞으로도 신경 써보시면 좋을거 같습니다!

고생하셨습니다 ~ 다음 step에서 뵙겠습니다 ~~ 👍 👍 👍
